src/blog/oilpan-pointer-compression.md

```diff
@@ -0,0 +1,247 @@
+
+## Implementation in Oilpan
+
+### Requirements
+
+So far, we talked about a trivial encoding scheme where the full pointer is formed by adding a base to an offset that is stored in a Member pointer.  The actually implemented scheme is unfortunately not as simple though, as Oilpan requires that Member can be assigned one of the following:
+
+1. A valid heap pointer to an object;
+2. The C++ `nullptr` (or similar);
+3. A sentinel value which must be known at compile time.  The sentinel value can e.g. be used to signal deleted values in hash tables that also support `nullptr` as entries.
+
+The problematic part around `nullptr` and a sentinel is the lack of explicit types to catch these on the caller side:
+
+```cpp
+void* ptr = member.get();
+if (ptr == nullptr) { /* ... * }
+```
+
+Since there’s no explicit type to store a possibly compressed `nullptr` value, an actual decompression is required to compare against the constant.
+
+Having this usage in mind, we were looking for a scheme that transparently handles case 1.-3. Since the compression and decompression sequence will be inlined everywhere Member is used, the following properties are also desirable:
+
+- Fast and compact instruction sequence to minimize icache misses.
+- Branchless instruction sequence to avoid using up branch predictors.
+
+Since it is expected that reads significantly outnumber writes, we allow for an asymmetric scheme where fast decompression is preferred.
+
+### Compression and decompression
+
+For brevity, this description only covers the final compression scheme used. See our [design doc](https://docs.google.com/document/d/1neGN8Jq-1JLrWK3cvwRIfrSjLgE0Srjv-uq7Ze38Iao) for more information on how we got there and the alternatives considered.
+
+The main idea for the scheme that is implemented as of today is to separate regular heap pointers from `nullptr` and sentinel by relying on alignment of the heap cage.  Essentially, the heap cage is allocated with alignment such that the least significant bit of the upper halfword is always set.  We denote the upper and lower half (32 bits each) as U<sub>31</sub>...U<sub>0</sub> and L<sub>31</sub>...L<sub>0</sub>, respectively.
+
+:::table-wrapper
+<!-- markdownlint-disable no-inline-html -->
+|              | upper half                               | lower half                                |
+| ------------ | ---------------------------------------: | ----------------------------------------: |
+| heap pointer | <tt>U<sub>31</sub>...U<sub>1</sub>1</tt> | <tt>L<sub>31</sub>...L<sub>2</sub>00</tt> |
```

Comment:
In the "lower half", please change
"<tt>L<sub>31</sub>...L<sub>2</sub>00</tt>"
to
"<tt>L<sub>31</sub>...L<sub>3</sub>000</tt>"